job-runner/chain/chain.go

```diff
@@ -97,103 +135,197 @@ func (c *chain) NextJobs(jobId string) proto.Jobs {
 }
 
 // JobIsReady returns whether or not a job is ready to run. A job is considered
-// ready to run if all of its previous jobs are complete. If any previous jobs
-// are not complete, the job is not ready to run.
+// ready to run if it is Pending or Stopped with some retry attempts remaining,
+// and all of its previous jobs are complete. If any previous jobs are not
+// complete, the job is not ready to run.
 func (c *chain) JobIsReady(jobId string) bool {
 	c.jcMux.RLock()
 	defer c.jcMux.RUnlock()
-	for _, job := range c.previousJobs(jobId) {
-		if job.State != proto.STATE_COMPLETE {
-			return false
+	return c.jobIsReady(jobId)
+}
+
+// RunnableJobs returns a list of all jobs that are ready to run. A job is
+// ready to run if all of its previous jobs are complete and it is Pending
+// or Stopped with some retries still remaining.
+func (c *chain) RunnableJobs() proto.Jobs {
+	runnableJobs := []proto.Job{}
+
+	// Loop through every job in the chain and check if it's ready to run. Only
+	// Pending + Stopped jobs may be ready to run.
+	for jobId, job := range c.jobChain.Jobs {
+		if !c.JobIsReady(jobId) {
+			continue
 		}
+
+		runnableJobs = append(runnableJobs, job)
 	}
-	return true
+
+	return runnableJobs
 }
 
-// IsDone returns two booleans - the first one indicates whether or not the
+// IsDoneRunning returns two booleans - the first one indicates whether or not the
 // chain is done, and the second one indicates whether or not the chain is
 // complete.
 //
-// A chain is done running if there are no more jobs in it that can run. This
-// can happen if all of the jobs in the chain or complete, or if some or all
-// of the jobs in the chain failed.
+// A chain is done running if there are no more jobs in it that can run or
+// are running. This can happen if all of the jobs in the chain are complete,
+// or if some or all of the jobs in the chain failed. Note that one failed job
+// does not mean the chain is done - there may still be pending jobs independent
+// of this failed job that can be run. Stopped jobs are treated the same as pending
+// jobs - they can be rerun (as they are when a suspended chain is resumed).
 //
 // A chain is complete if every job in it completed successfully.
-func (c *chain) IsDone() (done bool, complete bool) {
+func (c *chain) IsDoneRunning() (done bool, complete bool) {
 	c.jcMux.RLock()
 	defer c.jcMux.RUnlock()
 
 	done = true
 	complete = true
 	pendingJobs := proto.Jobs{}
 
+	// If any jobs are still running, the chain isn't done or complete.
+	for _, job := range c.jobChain.Jobs {
```

Comment:
This block is not necessary, redundant. The same check and return is done in the next block:
```go
        case proto.STATE_RUNNING:
            // We should never reach this case because we previously check for running jobs.
            return false, false
```